Optimize CI workflow.

player-03 opened the pull request:
The `-eval` option means most of the CI tasks no longer have to wait for all the others. If we run more of them in parallel, users will be able to get much quicker feedback on their pull requests.

player-03 committed:
CI: run android-ndll on Linux.

This way, it doesn't have to wait for the slow macos-ndll step.

player-03 committed:
CI: remove macos-ndll dependency from ios-ndll.

We want as few steps as possible to have to wait for macos-ndll. There's no way around package-haxelib, but `-eval` gives us a much better option for ios-ndll.

player-03 committed:
CI: remove linux-ndll dependency from android-ndll.

Not required, but it'll make the workflow summary neater.

player-03 committed:
CI: split `brew bundle` into its own step.

I'm almost sure Homebrew is what makes macos-ndll so slow, but this will help confirm it.

player-03 committed:
Rewrite `PlatformTarget.execute()`.

This function attempts to skip steps if the corresponding function was not overridden. However, in `-eval` mode, it may incorrectly detect the `@ignore` metadata on subclass functions, causing it to always skip.

Running an empty function is not a problem, but we wouldn't want to log that step if nothing was being done. Therefore I switched the check to only apply to the log call.

For one reason or another, the "watch" command can't be skipped, so I left it alone.

player-03 committed:
CI: run android-samples on Linux.

If switching to Linux is what made android-ndll so much faster, we should do the same for android-samples.

player-03 committed:
CI: give each sample its own step.

player-03 committed:
CI: run android-samples right after android-ndll.

This removes the need to wait on macos-ndll.

player-03 committed:
CI: update step names.

player-03 committed:
CI: remove dependencies from more tasks.

air-samples and neko-samples run on Haxe 3, which didn't have Eval mode, so they need to wait for the host platform's ndll.

Everything else only needs the ndll that will be included in the app (if any). Now we have some samples that can start immediately!

player-03 committed:
CI: install format before using `lime -eval`.

Maybe this requirement could be removed in the future? Not sure why `lime setup` would even need it.

player-03 committed:
CI: also install hxp.

player-03 committed:
CI: fix dependency errors.

Some of these make sense, but I have no idea why air-samples is complaining about hxcpp.

I also don't understand why macos-samples needs the submodules (or perhaps why linux-samples doesn't), but `-eval` should get around that.

player-03 committed:
CI: fix missing submodules.

For air-samples, this means downloading the submodules so that `lime rebuild tools` can work.

For windows-samples, we can use `-eval` to skip `rebuild tools` entirely.

player-03 commented:
At this point, it's possible to complete everything else while waiting for the `brew bundle` step of macos-ndll.

![Tasks](https://github.com/openfl/lime/assets/433823/8ef5b547-3288-4464-b39e-6eeb4ab5d7ab)

player-03 committed:
CI: switch to macos-12.

Apparently Homebrew dropped support for macOS 11 sometime last month, which may line up with when it started being extremely slow. Worth a shot!

player-03 commented:
Credit to @Apprentice-Alchemist for figuring out why Homebrew was so slow.

player-03 commented:
Workflow succeeded in 25 minutes, compared to ~1h45m before this PR.

joshtynjala commented:
Great to see the CI finishing faster! I'm concerned about one thing that you changed, though. The removal of this dependency from the samples builds:

```yml
needs: package-haxelib
```

I made the samples builds rely on package-haxelib on purpose. The idea was to ensure that the Haxelib .zip bundle that we're generating is valid, and can build projects for all supported targets.

player-03 commented:
I see where you're coming from. On the one hand, with the brew issue fixed (for now), going back to the package-haxelib bottleneck shouldn't slow things down _too_ much. On the other hand, it has been packaging correctly for years, it would throw an error if any of the pieces were missing, and we're testing the pieces individually. I don't see how it could start going wrong now.

How about this as a compromise? We test each target as soon as the required ndlls become available, saving time. Later when package-haxelib finishes, we test its output by downloading the bundle on Windows/Mac/Linux, and building a sample in HashLink. (HashLink because all we're trying to do is prove that the bundled version of Lime works.)

joshtynjala commented:
> How about this as a compromise? We test each target as soon as the required ndlls become available, saving time. Later when package-haxelib finishes, we test its output by downloading the bundle on Windows/Mac/Linux, and building a sample in HashLink. (HashLink because all we're trying to do is prove that the bundled version of Lime works.)

Works for me!

player-03 committed:
CI: after creating an ndll, immediately test it.

For instance, we're merging "android-ndll" with "android-samples" and merging "windows-ndll" with "windows-samples." There's no need to set up the same environment twice.

player-03 committed:
CI: validate package-haxelib output.

We need to make sure that lime-haxelib.zip is enough, on its own, to run Lime. And it needs to work on all three of Windows, Mac, and Linux.

Since this code will run at the very end of the workflow, not in parallel with anything, it's important to keep it quick. HashLink, Neko, and Flash are good candidates, but of those three, only HashLink is maintained.

player-03 committed:
CI: use single quotes.

player-03 committed:
CI: fix air-samples Haxe version.

player-03 committed:
CI: convert version numbers to variables.

player-03 committed:
CI: also run neko-samples after package-haxelib.

It's quick too, might as well test package-haxelib a bit more thoroughly.

player-03 committed:
CI: use `-eval` to run air-samples.

This allows it to run immediately, rather than having to wait for Windows. Plus, it makes the workflow summary clearly legible again.

The downside is that `-eval` wasn't available in Haxe 3, so we can no longer test that. However, I just added more Neko+Haxe 3 tests, so I think it balances out.

player-03 committed:
CI: use job outputs instead of variables.

player-03 committed:
CI: create dummy step for init job.

player-03 committed:
CI: try another way to get variables.

player-03 commented:
Ok, the variables gave me a fair bit of trouble, and I think the 25-minute run might have been a fluke, but it's working.

![Workflow](https://github.com/openfl/lime/assets/433823/60667d79-d6fb-45ef-b64d-95b2ff07bdb1)

player-03 commented:
Hmm, some of the Neko runs take a while to download the artifact. Maybe there are just too many in parallel. Could scale it back, I guess.

![Workflow](https://github.com/openfl/lime/assets/433823/8f82a66d-ccd0-49c1-b55b-439e5a4bac4a)

player-03 committed:
Reduce how many jobs compete for a download.

The post-package-haxelib jobs all need to download its output, and some of them (seemingly at random) spend an extra ~6m waiting for it.

player-03 commented:
Ugh, reducing the number of jobs didn't even help. All of the time saved was because the iOS builds happened to go a bit faster.

player-03 committed:
CI: skip HelloWorld sample on iOS.

The ios job is currently the biggest bottleneck, and this sample is the least informative of the three.

player-03 committed:
CI: skip HelloWorld sample on Mac.

Because apparently Mac can be the bottleneck sometimes.

player-03 commented:
As an experiment, let's try going back to hard-coded `runs-on` values. It's possible GitHub optimizes based on knowing in advance which machines to use, and that's how we got that 25-minute run.

player-03 committed:
CI: try hard-coding the `runs-on` values again.

player-03 committed:
CI: hard-code Haxe versions for neko-samples.

I guess you can use job output variables there, but not environment ones. But the environment variable is nicer for everything else, so we'll just not use a variable here.

player-03 commented:
It's definitely faster with hard-coded machines to run on. Still not 25 minutes, but I think that was only possible because we didn't test lime-haxelib.zip. @joshtynjala, does this look good to merge?
